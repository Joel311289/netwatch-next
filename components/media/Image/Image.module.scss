.wrapper {
  overflow: hidden;
  position: relative;

  &::before {
    content: '';
    display: block;
  }
}

.image {
  @include background-image;
  @include absolute(top 0 left 0 right 0 bottom 0);
}
