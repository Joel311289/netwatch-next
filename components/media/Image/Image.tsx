import { ElementChildren, ElementHTML, ElementSkeleton, Rectangle } from '@/types';
import { withSkeleton } from '@/hoc/withSkeleton';
import { useSizeRatio } from '@/hooks/useSizeRatio';
import { classes } from '@/utils/helpers';
import styles from '@/components/media/Image/Image.module.scss';

export type ImageProps = typeof defaultProps &
  ElementHTML &
  ElementChildren<JSX.Element> &
  ElementSkeleton &
  Rectangle & {
    src: string;
    ratio?: number;
    lazy?: boolean;
    alt?: string;
  };

const defaultProps = {};

const Image = ({ className, children, src, width, height, ratio, lazy, alt }: ImageProps) => {
  const sizes = useSizeRatio({ width, height, ratio });

  return (
    <div role="img" className={classes(styles.wrapper)}>
      <div
        aria-label={alt || 'image'}
        className={classes(styles.image, className)}
        {...(lazy && { 'data-background': src })}>
        {children}
      </div>

      <style jsx>{`
        .${styles.wrapper} {
          width: ${sizes.width || 'auto'};
          height: ${sizes.height || 'auto'};
          min-width: ${sizes.width || 'auto'};
        }
        .${styles.wrapper}::before {
          padding-top: ${ratio ? `${100 * ratio}%` : '0'};
        }

        .${styles.image} {
          ${!lazy && `background-image: ${src.toBackgroundImageUrl()}`};
        }
      `}</style>
    </div>
  );
};

Image.defaultProps = defaultProps;

const ImageWithSkeleton = withSkeleton<ImageProps>(Image);

export default ImageWithSkeleton;
