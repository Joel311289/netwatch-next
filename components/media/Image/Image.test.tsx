import { render, screen } from '@testing-library/react';
import Image from '@/components/media/Image/Image';

describe('Tests Image component', () => {
  test('renders component correctly', () => {
    const { container } = render(
      <Image alt="img" src="https://image.tmdb.org/t/p/w500/4pCSBPHUPia93rppHF3UX4cLQ9M.jpg" />
    );

    expect(container).toMatchSnapshot();
  });

  test('renders with ratio correctly', () => {
    const width = 200;
    const ratio = 1.5;
    render(
      <Image
        alt="img"
        src="https://image.tmdb.org/t/p/w500/4pCSBPHUPia93rppHF3UX4cLQ9M.jpg"
        width={width}
        ratio={ratio}
      />
    );

    expect(screen.getByRole('img')).toHaveStyle(`width: ${width}px`);
    expect(screen.getByRole('img')).toHaveStyle(`height: ${width * ratio}px`);
  });
});
