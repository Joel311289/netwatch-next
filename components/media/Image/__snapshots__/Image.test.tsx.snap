// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Image component renders component correctly 1`] = `
<div>
  <div
    class="wrapper"
    role="img"
  >
    <div
      aria-label="img"
      class="image"
    />
    <style>
      
        .wrapper {
          width: auto;
          height: auto;
          min-width: auto;
        }
        .wrapper::before {
          padding-top: 0;
        }

        .image {
          background-image: url(https://image.tmdb.org/t/p/w500/4pCSBPHUPia93rppHF3UX4cLQ9M.jpg);
        }
      
    </style>
  </div>
</div>
`;
